bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ServiceCaller.java

```diff
@@ -160,6 +165,124 @@ public static <S> boolean callOnce(Class<?> caller, Class<S> serviceType, String
 		}).orElse(Boolean.FALSE);
 	}
 
+	/**
+	 * Not thread save.
+	 * <p>
+	 * Service registrations are obtained upon creation, service instances are get
+	 * on the first request (i.e. the first instance when {@link #first()} or
+	 * {@link #all()} is called. Services that are unregistered in the meantime are
+	 * discarded and newly registered services are ignored.
+	 * </p>
+	 * 
+	 * @param <S> the service type for this caller
+	 * @since 3.17
+	 */
+	public interface ServiceUsageScope<S> extends AutoCloseable { // TODO:? , Supplier<S>
+
+		/**
+		 * Gets the first available service implementation.
+		 */
+		Optional<S> first();
```

Comment:
first can be default implemented as
```suggestion
		default Optional<S> first() {
		      return all().findFirst();		
		}
```

Reply:
True, but realistically there will only ever be one implementation of the interface so there is no real value in using a default method.

Reply:
Should this be a public final class with a package visible constructor so nothing else can construct it?  That makes it clear there are no other implementations.

Reply:
> Should this be a public final class with a package accessible constructor so nothing else can construct it? That makes it clear there are no other implementations.

Done.